File: leetcode/34/solution.cpp
#include "iostream"
#include "vector"

using namespace std;

class Solution {
public:
  vector<int> searchRange(vector<int> &nums, int target) {
    int l = nums.size();
    int i = l / 2;
    int min = 0, max = l - 1;
    int begin = -1, end = -1;

    if (l == 0 || nums.front() > target || nums.back() < target) {
      return {-1, -1};
    }

    while (true) {
      if (min == max) {
        break;
      }
      if (nums[i] == target) {
        int begin = target, end = target;
        while (0 <= begin - 1 && nums[begin - 1] == target) {
          begin--;
        }
        while (end + 1 < l && nums[end + 1] == target) {
          end++;
        }
        break;
      }

      if (nums[i] < target) {
        i += (l - min) / 2;
        min = i;
      } else {
        i -= (l - i) / 2;
        max = i;
      }
    }

    return {begin, end};
  }
};

int main() {
  // Solution s = Solution();
  // vector<int> v = {5, 7, 7, 8, 8, 10};
  int expected[2] = {3, 4};

  for (auto item : expected) {
    std::cout << item << " ";
  }
  if (expected == expected) {
    std::cout << "correct" << std::endl;
  } else {
    std::cout << "wrong" << std::endl;
  }
  return 0;
}

File: leetcode/34/solution.py
"""
https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/submissions/
34. Find First and Last Position of Element in Sorted Array
"""
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        answer = [-1, -1]
        if len(nums) == 0 or target < nums[0] or nums[-1] < target:
            return answer

        l, r = 0, len(nums) - 1
        while l < r:
            m = (l + r) // 2
            if target <= nums[m]:
                r = m
            else:  # nums[m] < target
                l = m + 1

        if nums[l] != target:
            return [-1, -1]

        answer[0] = l
        r = len(nums) - 1
        while l < r:
            m = (l + r) // 2
            if nums[m] <= target:
                l = m + 1
            else:  # target < nums[m]
                r = m

        answer[1] = r if nums[r] == target else r - 1
        return answer


"""
# Optimized version:
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0 or target < nums[0]  or nums[-1] < target:
            return [-1, -1]

        l = bisect_left(nums, target)
        if nums[l] != target:
           return [-1, -1] 
        r = bisect_right(nums, target)
        
        return [l, r - 1] if r == len(nums) or nums[r] != target else [l, r]
"""
